#2: Implemented lifecycle of an EC2 instance

tomuben opened the pull request:
fixes #2

tomuben committed:
#2: Implemented launch of an EC2 instance

tomuben committed:
Added git to dev-dependencies

tomuben committed:
Fixed integration test

tomuben committed:
1. Moved key-file parameters outside of aws_options.py
2. moved method STACK_NAME into class CloudFormationStack

tomuben committed:
Refactored run_setup_ec2.py and added tests for run_setup_ec2.py and key_file_manager.py

tkilias commented on exasol_script_languages_developer_sandbox/lib/random_string_generator.py:
```diff
@@ -0,0 +1,13 @@
+    Creates a random string of given length.
+    """
+    # choose from all lowercase letter
+    letters = string.ascii_lowercase
```
Maybe use the following base64(hash(uuid))[:length]

tkilias commented on exasol_script_languages_developer_sandbox/lib/run_setup_ec2.py:
```diff
@@ -0,0 +1,54 @@
+from exasol_script_languages_developer_sandbox.lib.key_file_manager import KeyFileManager
+
+
+def unpack_ec2_instance_descript(ec2_instance_description: Any) -> Tuple[str, str]:
```
```suggestion
def unpack_ec2_instance_description(ec2_instance_description: Any) -> Tuple[str, str]:
```

tkilias commented on exasol_script_languages_developer_sandbox/lib/run_setup_ec2.py:
```diff
@@ -0,0 +1,54 @@
+    return ec2_instance_description["State"]["Name"], ec2_instance_description["PublicDnsName"]
+
+
+def execute_setup_ec2(aws_access: AwsAccess,
```
I think, the name doesn't fit

tomuben replied:
renamed it to run_lifecycle_for_ec2

tkilias commented on exasol_script_languages_developer_sandbox/templates/ec2_cloudformation.jinja.yaml:
```diff
@@ -0,0 +1,24 @@
+Resources:
+  Ec2SecurityGroup:
```
Add Exasol tags

tkilias commented on exasol_script_languages_developer_sandbox/lib/deployer.py:
```diff
@@ -0,0 +1,212 @@
+        description = "Created by AWS CLI at {0} UTC".format(now)
+
+        # Each changeset will get a unique name based on time
+        changeset_name = self.changeset_prefix + str(int(time.time()))
```
I guess, we only need one changeset, such that time is unique enough

tomuben replied:
This is the code I copied from AWS CLI. I would prefer to keep it as is.

tkilias replied:
Ah missed that. Then it's ok.

tomuben committed:
1. Added tags to AWS resources
2. Changed implementation of random_string_generator.py
3. renamed function execute_setup_ec2() => run_lifecycle_for_ec2()
4. minor changes

tkilias commented on exasol_script_languages_developer_sandbox/lib/cf_stack.py:
```diff
@@ -0,0 +1,48 @@
+from exasol_script_languages_developer_sandbox.lib.render_template import render_template
+
+
+class CloudformationStack:
```
Does a ContexManager makes sense here?

tomuben replied:
Done

tkilias commented on exasol_script_languages_developer_sandbox/lib/key_file_manager.py:
```diff
@@ -0,0 +1,40 @@
+
+
+class KeyFileManager:
+    def __init__(self, aws_access: AwsAccess,
```
Does a ContexManager makes sense here?

tomuben replied:
Done

tomuben committed:
converted CloudformationStack into a ContextManager

tomuben committed:
converted KeyFileManager into a ContextManager

tkilias commented on test/test_key_file_manager.py:
```diff
@@ -27,16 +26,17 @@ def test_generated_key():
-    # Now call close. After that we expect that the key has been removed from AWS and the temporary file was removed.
-    km.close()
-    aws_access_mock.delete_ec2_key_pair.assert_called_once_with(key_name=km.key_name)
+        # Now call close. After that we expect that the key has been removed from AWS and the temporary file was removed.
```
```suggestion
    # Now call close. After that we expect that the key has been removed from AWS and the temporary file was removed.
```

tomuben replied:
Done

tkilias reviewed (changes requested):
Some questions and thoughts. Should we use aws localstack for testing the aws_access class. And can we refactor things, such that we have the context manager for keys and cloud formation, but on the other hand, use creation and closing idependently as CLI. Means the create CLI stores everything necessary on disk and close loads this and then closes everything. Reason is, I need the exact functionalities for handling ec2 instances also for GPU tests and development.
Another question, could we add the Key creation into the cloud formation template.

tomuben commented:
> Another question, could we add the Key creation into the cloud formation template.

It is possible, see [here](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-keypair.html), but I there are a few reasons to not do this:
1. For some cases (maybe debugging/tests) we might want to use an existing KeyPair. In order to implement that we would have to assemble the cloudformation yaml file dynamically, which makes the code more difficult.
2. We still need to get access to the private key of the Keypair. I assume the best way would be to write the private key to a temporary file (as we do now). Then the management of the lifecycle of this file would become independent of the lifecycle of the KeyPair (which is managed within the lifecycle of the cloudformation stack)
3. The code to get access to the private key is not much easier: First one has to find the Key-Id, and after that use AWS SSM to get the value of the private key.

tomuben commented:
>And can we refactor things, such that we have the context manager for keys and cloud formation, but on the other hand, use creation and closing idependently as CLI. Means the create CLI stores everything necessary on disk and close loads this and then closes everything. Reason is, I need the exact functionalities for handling ec2 instances also for GPU tests and development.

Of, course, I am happy if the code will be reused. But the question is if you want to use it from within Python code, or via CLI. Let's discuss offline.

tomuben committed:
Fixed comment from review

tomuben committed:
Add a localstack test

tomuben committed:
Added more tests to localstack test

tomuben commented:
> Should we use aws localstack for testing the aws_access class.

I added new tests which use `localstack`. However I realized that localstack simply mocks many things. It's not possible to test error scenarios (e.g. verify that a client error is thrown when a cloudformation stack is being accessed after it has been deleted).

tomuben committed:
Removed ContextManager behavior from class KeyFileManager

tomuben committed:
Use separate class KeyFileManagerContextManager

tomuben committed:
Added tests for serialization/deserialization of classes KeyFileManager()/CloudformationStack()

tomuben committed:
fixed tests

tkilias commented on test/test_serialization.py:
```diff
@@ -0,0 +1,79 @@
+    aws_access = AwsLocalStackAccess(None)
+    key_file_manager = KeyFileManager(aws_access, None, None)
+    key_file_manager.create_key_if_needed()
+    q.put(key_file_manager.key_name)
+    q.put(key_file_manager.key_file_location)
+    with open(tmp_location, "wb") as f:
+        pickle.dump(key_file_manager, f)
```
first pickle then put, otherwise a race condition is possible

tomuben replied:
I do a join in line 32, so I believe there is no race condition.

tkilias replied:
you are right

tkilias commented on test/test_run_lifecycle_for_ec2.py:
```diff
@@ -0,0 +1,46 @@
+    aws_access_mock.describe_instance.side_effect = instances_states
+    res_gen = run_lifecycle_for_ec2(aws_access_mock, "test_key_file_loc", "test_key")
+    res = next(res_gen)
+    assert not aws_access_mock.create_new_ec2_key_pair.called
+    assert aws_access_mock.upload_cloudformation_stack.called
+    assert res == ("pending", "public_host", "abc", "test_key_file_loc")
+    res = next(res_gen)
+    assert res == ("pending", "public_host", "abc", "test_key_file_loc")
+    res = next(res_gen)
+    assert res == ("running", "public_host", "abc", "test_key_file_loc")
+
+    #Check cleanup
+    assert not aws_access_mock.delete_stack.called
+    next(res_gen)
+    assert aws_access_mock.delete_stack.called
```
I think, here we can't avoid the multiple asserts, because you are testing a sequence of events

tomuben replied:
Also, the test is really fast.

tkilias replied:
ok, that case we can leave it

tkilias commented on test/test_key_file_manager.py:
```diff
@@ -0,0 +1,42 @@
+    with open(tmp_file, "w") as f:
+        f.write("secret")
+    with KeyFileManagerContextManager(KeyFileManager(aws_access_mock, "test_key", str(tmp_file))) as km:
+        assert km.key_name == "test_key"
+        assert km.key_file_location == str(tmp_file)
+        assert not km._remove_key_on_close
+    assert not aws_access_mock.delete_ec2_key_pair.called
+    assert os.path.exists(str(tmp_file))
```
split test to avoid multiple assert

tomuben replied:
Hm, do you think this is a problem in that case?
I understood multiple asserts are mainly an issue for long running tests. Here the tests run within seconds...

tkilias replied:
ok, in that case not so important

tkilias commented on exasol_script_languages_developer_sandbox/lib/deployer.py:
```diff
@@ -0,0 +1,212 @@
+# Copyright 2012-2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
hm, we have this now already in two projects. I am thinking, about if we should move this to its own package

tomuben replied:
Hm, maybe something like boto3-extionsion, or aws-python-access

tomuben replied:
I created a ticket for this: https://github.com/exasol/script-languages-developer-sandbox/issues/9

tomuben committed:
1. Use base64 for getting random string based on UUID
2. For EC2 Key use full uuid string
3. For Cloudformation stack check if stack name exists

tomuben committed:
Fixed test_run_lifecycle_for_ec2.py

tkilias commented on exasol_script_languages_developer_sandbox/lib/cf_stack.py:
```diff
@@ -25,15 +29,28 @@ def _generate_stack_name() -> str:
         so that in theory multiple instances can be created.
         """
-        return f"EC2-SLC-DEV-SANDBOX-{get_random_str(5)}"
+        return f"EC2-SLC-DEV-SANDBOX-{get_random_str_of_length_n(5)}"
```
prefix parameter

tomuben replied:
Done

tkilias commented on exasol_script_languages_developer_sandbox/lib/random_string_generator.py:
```diff
@@ -1,8 +1,17 @@
+def get_random_str() -> str:
+    """
+    Creates a random string of length 22.
+    Uses uuid.uuid4() as base and replaces "+","/" with "A"/"B" and removes any filling characters ("=").
```
ok, I probably had removed + and  /, but this way you have a constant length. Maybe, add explanation, why base64

tomuben replied:
Done

tomuben committed:
Fixed tests

tomuben committed:
Use same version for localstack docker image as Python package

tomuben committed:
Added documentation

tomuben committed:
Fixed test_serialization.py

tkilias commented on exasol_script_languages_developer_sandbox/lib/cf_stack.py:
```diff
@@ -17,19 +15,23 @@ class CloudformationStack:
         self._ec2_key_name = ec2_key_name
         self._user_name = user_name
+        if stack_prefix is None:
+            self._stack_prefix = "EC2-SLC-DEV-SANDBOX-"
```
you should be able to put the default directly to the parameter. This is only an issue if the default has a mutable type, strings shouldn't be mutable.

tomuben replied:
as I pass through the prefix in function `run_lifecycle_for_ec2`, I would have to add a check there, too. Which makes it even more complicated

tkilias replied:
ok, you are right

tkilias commented on exasol_script_languages_developer_sandbox/lib/random_string_generator.py:
```diff
@@ -0,0 +1,19 @@
+    Creates a random string of length 22.
+    Uses uuid.uuid4() as input for Base64,
+    which replaces "+","/" with "A"/"B" and removes any filling characters ("=").
+    We use Base64 (and not for example hex) as it uses all alphanumerical characters.
```
```suggestion
    We use Base64 (and not for example hex) as it uses all alphanumerical characters and
    with this increase the possibilties per character and the decrease probability for a collision
    for a given length. 
```

tkilias commented on test/test_serialization.py:
```diff
@@ -0,0 +1,82 @@
+import multiprocessing as mp
+
+
+def create_key_pair_and_serialize(tmp_location: Path, q: mp.Queue):
```
wrap content maybe in try except with print exception, because otherwise you don't get any hint what is going wrong

tkilias commented on test/test_serialization.py:
```diff
@@ -0,0 +1,82 @@
+    assert os.path.exists(restored_key_file_location) is False
+
+
+def create_cloudformation_stack_and_serialize(tmp_location_key_manager: Path, tmp_location_cloudformation: Path,
```
wrap content maybe in try except with print exception, because otherwise you don't get any hint what is going wrong

tomuben committed:
Minor fixes from review

tkilias commented on test/test_serialization.py:
```diff
@@ -10,13 +11,17 @@
+            pickle.dump(key_file_manager, f)
+    except Exception as e:
+        sys.stderr.write("failed to execute create_key_pair_and_serialize\n")
+        sys.stderr.write(str(e))
```
I suggest traceback.print_exc()

tkilias commented on test/test_serialization.py:
```diff
@@ -45,17 +49,21 @@ def test_keypair_manager_with_local_stack(tmp_path, local_stack):
+        q.put(cloudformation.get_ec2_instance_id())
+    except Exception as e:
+        sys.stderr.write("failed to execute create_cloudformation_stack_and_serialize\n")
+        sys.stderr.write(str(e))
```
I suggest traceback.print_exc()

tomuben committed:
Minor fixes from review
